### PR TITLE
[web][api][SIMS #540] Separate Institution Details and Institution User Profile

### PR DESCRIPTION
**Pull request by dheepak-aot**
Separate Institution Details and Institution User Profile Information #540

### CODE REVIEW

**Comment by abschwenker** [comment]
Please have a look at the other tests failing as well

File: sources/packages/api/src/services/user/user.service.ts [hunk 1]
```diff
@@ -51,6 +51,15 @@ export class UserService extends DataModelService<User> {
       .execute();
   }
 
+  async updateUserEmail(userId: number, email: string) {
```
Comment (ann-aot) [inline sources/packages/api/src/services/user/user.service.ts:54]: can u add a Promise with the return Type

File: sources/packages/api/src/route-controllers/user/models/institution-user.dto.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { IsNotEmpty } from "class-validator";
+
+export class InstitutionUserDto {
+  @IsNotEmpty()
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/user/models/institution-user.dto.ts:4]: The class validators should be used to validate input DTOs, they have effect on response DTOs.
Please remove the class validators.

File: sources/packages/api/src/route-controllers/user/user.controller.ts [hunk 3]
```diff
@@ -120,4 +121,16 @@ export class UserController extends BaseController {
       throw error;
     }
   }
+
+  @Get("/institutionUser")
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/user/user.controller.ts:125]: IMO, the information being returned here is not institution-specific. I would recommend that this endpoint would be generic GET users.

File: sources/packages/api/src/route-controllers/user/user.controller.ts [hunk 4]
```diff
@@ -120,4 +121,16 @@ export class UserController extends BaseController {
...
+  async institutionDetail(
+    @UserToken() userToken: IUserToken,
+  ): Promise<InstitutionUserDto> {
+    const user = await this.service.getActiveUser(userToken.userName);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/user/user.controller.ts:129]: Please add a check if the user is present, if not, throw an HTTP error exception.

File: sources/packages/api/src/route-controllers/user/user.controller.ts [hunk 5]
```diff
@@ -120,4 +121,16 @@ export class UserController extends BaseController {
...
+    const institutionUser = new InstitutionUserDto();
+    institutionUser.userEmail = user?.email;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/user/user.controller.ts:131]: At this point, the user must have a value and there is no need for a null check.

File: sources/packages/api/src/services/user/user.service.ts (hunk 1 above)
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/user/user.service.ts:54]: Just a recommendation, this code could be simplified as below:
```
return this.repo.update({id: userId}, { email: email });
```

File: sources/packages/web/src/router/InstitutionRoutes.ts [hunk 6]
```diff
@@ -213,6 +214,17 @@ export const institutionRoutes: Array<RouteRecordRaw> = [
...
       },
+      {
+        path: AppRoutes.InstitutionUserProfile,
+        name: InstitutionRoutesConst.INSTITUTION_USER_PROFILE,
+        components: {
+          default: InstitutionUserProfile,
+          sidebar: InstitutionHomeSideBar,
+        },
+        meta: {
+          clientType: ClientIdType.INSTITUTION,
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/router/InstitutionRoutes.ts:225]: I am not sure if we need explicitly set the user types also. @ann-aot, can you please confirm it? It would be something as below:
```
userTypes: [
            InstitutionUserTypes.admin,
            InstitutionUserTypes.locationManager,
            InstitutionUserTypes.user,
          ],
```

File: sources/packages/web/src/services/UserService.ts [hunk 7]
```diff
@@ -37,4 +38,10 @@ export class UserService {
   async checkActiveUser(authHeader?: any): Promise<boolean> {
     return ApiClient.User.checkActiveUser(authHeader);
   }
+
+  async getInstitutionUser(
+    authHeader?: any,
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/UserService.ts:43]: The auth header is being passed as a parameter in exceptional cases when the authentication token is not available globally yet. Just a few methods need this parameter. If this one does not need it, please consider removing it.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    // Hooks
+    const toast = useToast();
+    const router = useRouter();
+    // Data-bind
+    const initialData = ref({});
+
+    const submitted = async (data: InstitutionDto) => {
+      let redirectHome = true;
+      try {
+        await InstitutionService.shared.updateInstitute(data);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:41]: My understanding is that this screen is only updating the user email. This method is updating much other information for the institution and I believe that it will erase a lot of data on the institution side.
Can you please check if makes sense to use this method? It seems to me that the information should be saved using a specific method.

Comment (andrewsignori-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue:41]: IMO, this is user-specific info and it should be in the user service.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      if (redirectHome) {
+        router.push({
+          name: InstitutionRoutesConst.INSTITUTION_DASHBOARD,
+        });
+      }
+    };
+
+    // Hooks
+    onMounted(async () => {
+      const bceidAccount = await UserService.shared.getInstitutionUser();
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:67]: This should be treated as our user data since it is in our DB. The use of the BCeID account could lead to misinterpretations. Could you please rename it to something like "userAccount", please?

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      if (bceidAccount) {
+        initialData.value = {
+          userFirstName: bceidAccount?.userFirstName,
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:70]: There is already a check in place on line 68 to ensure that the variable is defined. Could you please remove the null check while accessing the properties?

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+          userLastName: bceidAccount?.userLastName,
+          userEmail: bceidAccount?.userEmail,
+        };
+      } else {
+        toast.add({
+          severity: "error",
+          summary: "BCeID Account error",
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:77]: IMO,  this is not related to BCeID account, can you please adjust the message.

**Review by andrewsignori-aot — CHANGES_REQUESTED** [review]
Please review the comments.
Is this commit a partial delivery of the ticket or is it supposed to deliver it entirely?
If it is a complete delivery, there are some missing acceptance criteria, IMO.

Once the institutionuserprofile form definition is completed it must be added to code repo, the same for any other form definition that was updated as part of this ticket.

There are some readonly components in the screen that do not need to be sent to the API and they could be configured as below.
image

File: sources/packages/web/src/router/InstitutionRoutes.ts (hunk 6 above)
Comment (ann-aot) [reply sources/packages/web/src/router/InstitutionRoutes.ts:225]: sorry @dheepak-aot  I mean, for institutions you will need to add the userType. as @andrewsignori-aot  said it is working for you since the student router does any restriction, we need to adding logic for the student router in future sprint

File: sources/packages/web/src/services/UserService.ts [hunk 12]
```diff
@@ -37,4 +38,17 @@ export class UserService {
...
+  ): Promise<InstitutionUserDetailsDto> {
+    return ApiClient.User.getinstitutionUser(authHeader);
+  }
+
+  async updateInstitutionUser(
+    data: InstitutionUserDetailsDto,
+    authHeader?: any,
+  ): Promise<void> {
+    return ApiClient.User.updateInstitutionUser(data, authHeader);
```
Comment (ann-aot) [inline sources/packages/web/src/services/UserService.ts:52]: promising void and you are returning a value

File: sources/packages/web/src/services/http/UserApi.ts [hunk 13]
```diff
@@ -56,4 +57,35 @@ export class UserApi extends HttpBaseClient {
...
+      throw error;
+    }
+  }
+
+  public async updateInstitutionUser(
+    data: InstitutionUserDetailsDto,
+    headers?: any,
+  ): Promise<void> {
+    try {
+      const response = await this.apiClient.patch(
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/UserApi.ts:81]: i think const response is not used inside the scope, you can remove it

File: sources/packages/web/src/views/institution/AppInstitution.vue [hunk 14]
```diff
@@ -1,5 +1,5 @@
 <template>
-  <v-app-bar dense flat app>
+  <v-app-bar dense flat app style="overflow:visible">
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/AppInstitution.vue:2]: can you share, why this style was added?

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        redirectHome = false;
+        toast.add({
+          severity: "error",
+          summary: "Unexpected error",
+          detail: "An error happened during the update process.",
+          life: 5000,
+        });
+      }
+
+      if (redirectHome) {
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:57]: IMO, I think you can avoid this check and add the redirect logic inside the try

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    onMounted(async () => {
+      const bceidAccount = await UserService.shared.getInstitutionUser();
+      if (bceidAccount) {
+        initialData.value = {
+          userFirstName: bceidAccount?.userFirstName,
+          userLastName: bceidAccount?.userLastName,
+          userEmail: bceidAccount?.userEmail,
+        };
+      } else {
+        toast.add({
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:74]: I think @andrewsignori-aot, has added utils for toast, we can use that from now on. correct me if I am wring @andrewsignori-aot

**Review by ann-aot — COMMENTED** [review]
Added some comments

File: sources/packages/api/src/route-controllers/user/user.controller.ts (hunk 3 above)
Comment (guru-aot) [reply sources/packages/api/src/route-controllers/user/user.controller.ts:131]: I guess we decided to have it as /institution. Right @dheepak-aot, @andrewsignori-aot @ann-aot  ?

Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/user/user.controller.ts:131]: Yes @guru-aot

**Review by guru-aot — COMMENTED** [review]
Please check on @ann-aot @andrewsignori-aot comments, I do second them on the comments. Nice overall work @dheepak-aot :)

File: sources/packages/web/src/views/institution/AppInstitution.vue (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/institution/AppInstitution.vue:2]: It is also not clear to me also.

File: sources/packages/web/src/views/institution/AppInstitution.vue [hunk 17]
```diff
@@ -88,10 +114,34 @@ export default {
...
+    const togleUserMenu = (event: any) => {
+      userOptionsMenuRef.value.toggle(event);
+    };
+
+    userMenuItems.value = [
+      {
+        label: "Notifications Settings",
+        icon: "pi pi-bell",
+        command: () => {
+          AppConfigService.shared.logout(ClientIdType.INSTITUTION);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/AppInstitution.vue:126]: The notifications settings should not point to the logout. I would recommend adding a placeholder page in the same way that we did for Students.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+    const submitted = async (data: InstitutionUserDetailsDto) => {
+      const institutionUserPersistDto: InstitutionUserPersistDto = {
+        userEmail: data.userEmail,
+      };
+      try {
+        await UserService.shared.updateInstitutionUser(
+          institutionUserPersistDto,
+        );
+        toast.success("Updated!", "Institution and User successfully updated!");
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/InstitutionUserProfile.vue:48]: The message is mentioning the institution, but I believe that only the user is being updated here, right?

File: sources/packages/web/src/views/institution/AppInstitution.vue (hunk 14 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/AppInstitution.vue:2]: The same style is used in the nav bar of student page. AppStudent.vue. without this style items in menu bar hide inside the nav bar.

Like this

image

Instead of 

image

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 16 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Toast utils updated

File: sources/packages/web/src/views/institution/AppInstitution.vue (hunk 17 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/AppInstitution.vue:126]: Agreed. Forgot to remove that action for notifications icon.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 18 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue:48]: Again a copy paste issue.  will fix it.

File: sources/packages/web/src/services/http/UserApi.ts (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/UserApi.ts]: Yes. The const has been removed

File: sources/packages/web/src/services/UserService.ts (hunk 12 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/UserService.ts]: return statement has been removed.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 15 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: As discussed the redirect logic is removed from condition.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Message modified.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Null check removed.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: const has been renamed.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    const initialData = ref({});
+
+    const submitted = async (data: InstitutionDto) => {
+      let redirectHome = true;
+      try {
+        await InstitutionService.shared.updateInstitute(data);
+        toast.add({
+          severity: "success",
+          summary: "Updated!",
+          detail: "Institution and User successfully updated!",
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Messages have been updated accordingly.

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Yes. Now user service is being used here.

File: sources/packages/web/src/services/http/UserApi.ts [hunk 20]
```diff
@@ -56,4 +57,19 @@ export class UserApi extends HttpBaseClient {
       throw error;
     }
   }
+
+  public async getinstitutionUser(
+    headers?: any,
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/UserApi.ts]: The headers argument has been removed

File: sources/packages/web/src/services/UserService.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/UserService.ts]: The headers argument has been removed

File: sources/packages/api/src/route-controllers/user/user.controller.ts (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/user/user.controller.ts]: Null check has been removed

File: sources/packages/api/src/route-controllers/user/user.controller.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/user/user.controller.ts:136]: User is validated and exception is thrown accordingly

File: sources/packages/api/src/route-controllers/user/models/institution-user.dto.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/user/models/institution-user.dto.ts]: Class validator used only for persist DTO

File: sources/packages/web/src/views/institution/InstitutionUserProfile.vue (hunk 18 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/InstitutionUserProfile.vue]: Fixed

File: sources/packages/web/src/views/institution/AppInstitution.vue (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/institution/AppInstitution.vue:2]: Yes, now I remembered. It is because we are using PrimeVue because Veutify was not ready during the time that the Student was updated.

**Review by andrewsignori-aot — APPROVED** [review]
Nice start 😉 👍

**Review by abschwenker — APPROVED** [review]
LGTM, thank you for reviewing the comments. Once everyone is approved please merge

**Review by ann-aot — APPROVED** [review]
Nice work @dheepak-aot  👍
